### PR TITLE
Feat: obj_controller refactor 4

### PR DESCRIPTION
**Pull request by felipeStouvenel**
# Feat
- Refactor of mouse_50.gml of obj_controller
- Improved logic and documented the code

### CODE REVIEW

File: objects/obj_controller/Mouse_50.gml [hunk 1]
```diff
@@ -75,35 +81,55 @@ if (menu=12) and (cooldown<=0) and (penitorium>0){
...
+                obj_ini.armour[c,e]="";
+                obj_ini.gear[c,e]="";
+                obj_ini.hp[c,e]=100;
+                obj_ini.chaos[c,e]=0;
+                obj_ini.experience[c,e]=0;
+                obj_ini.mobi[c,e]="";
+                obj_ini.age[c,e]=0;
+                obj_ini.spe[c,e]="";
+                obj_ini.god[c,e]=0;
+                diplo_char=c;
```
Comment (OH296) [inline objects/obj_controller/Mouse_50.gml:109]: there is now a generic function for this which should be favoured when the inevitable merge conflict emerges

File: objects/obj_controller/Mouse_50.gml [hunk 2]
```diff
@@ -112,96 +138,119 @@ if (menu=12) and (cooldown<=0) and (penitorium>0){
...
+                var pop=instance_create(0,0,obj_popup);
+                pop.type=9;
+                cooldown=8;
             }
         }
-        if (mouse_x>=xx+780) and (mouse_y>=yy+789) and (mouse_x<xx+894) and (mouse_y<yy+804) and (cooldown<=0){// Destroy
-            var fun;fun=floor(random(100))+1;
+        // Destroy
+        if (mouse_x>=xx+780) and (mouse_y>=yy+789) and (mouse_x<xx+894) and (mouse_y<yy+804) and (cooldown<=0){
+            var fun=floor(random(100))+1;
```
Comment (OH296) [inline objects/obj_controller/Mouse_50.gml:213]: irandom instead? totatlly optional

File: objects/obj_controller/Mouse_50.gml [hunk 3]
```diff
@@ -112,96 +138,119 @@ if (menu=12) and (cooldown<=0) and (penitorium>0){
...
-                pop.type=9;cooldown=8;
+                var pop=instance_create(0,0,obj_popup);
+                pop.type=9;
+                cooldown=8;
             }
         }
-        if (mouse_x>=xx+780) and (mouse_y>=yy+789) and (mouse_x<xx+894) and (mouse_y<yy+804) and (cooldown<=0){// Destroy
-            var fun;fun=floor(random(100))+1;
+        // Destroy
+        if (mouse_x>=xx+780) and (mouse_y>=yy+789) and (mouse_x<xx+894) and (mouse_y<yy+804) and (cooldown<=0){
```
Comment (OH296) [inline objects/obj_controller/Mouse_50.gml:212]: in general i think we should use the point_in_rectangle() function gamemaker ships with for these

File: objects/obj_controller/Mouse_50.gml [hunk 4]
```diff
@@ -112,96 +138,119 @@ if (menu=12) and (cooldown<=0) and (penitorium>0){
...
-            var i;i=menu_artifact;
+            var i=menu_artifact;
 
-            if (menu_artifact=fest_display) then fest_display=0;
+            if (menu_artifact==fest_display) then fest_display=0;
 
             if (string_count("Daemon",obj_ini.artifact_tags[i])>0){
                 if (obj_ini.artifact_sid[i]>=500){
-                    var oops;oops=floor(random(100))+1;
+                    var oops=floor(random(100))+1;
```
Comment (OH296) [inline objects/obj_controller/Mouse_50.gml:221]: oops works if your reading the code in context but could be an overall clearer name.
also irandom? again idk

File: objects/obj_controller/Mouse_50.gml [hunk 5]
```diff
@@ -210,47 +259,31 @@ if (menu=13) and (cooldown<=0) and (artifacts>0){
...
-    if (stc_wargear_un+stc_vehicles_un+stc_ships_un=0) then draw_set_alpha(0.5);
-    draw_text(xx+761,yy+467,"Gift");draw_set_alpha(1);
-    */
-
-
-    if (mouse_x>=xx+733) and (mouse_y>=yy+466) and (mouse_x<xx+790) and (mouse_y<yy+486) and (cooldown<=0){// Gift STC Fragment
+// ** Armamentorium STC fragments **
+if (menu==14) and (cooldown<=0){
+    // Gift STC Fragment
+    if (mouse_x>=xx+733) and (mouse_y>=yy+466) and (mouse_x<xx+790) and (mouse_y<yy+486) and (cooldown<=0){
```
Comment (OH296) [inline objects/obj_controller/Mouse_50.gml:266]: point_in_rectangle maybe?

File: objects/obj_controller/Mouse_50.gml [hunk 6]
```diff
@@ -259,251 +292,370 @@ if (menu=14) and (cooldown<=0){
...
+if (menu==15) and (cooldown<=0){
     if (mouse_x>=xx+748) and (mouse_x<xx+772){
-        if (mouse_y>=yy+355) and (mouse_y<yy+373) and (recruiting<5) and (obj_controller.gene_seed>0) and (obj_ini.doomed=0) and (string_count("|",obj_controller.recruiting_worlds)>0) and (obj_controller.penitent=0){cooldown=8000;recruiting+=1;income_recruiting-=2*(string_count("|",obj_controller.recruiting_worlds));scr_income();}
-
-        if (mouse_y>=yy+395) and (mouse_y<yy+413) and (training_apothecary<6){cooldown=8000;training_apothecary+=1;scr_income();}
-        if (mouse_y>=yy+415) and (mouse_y<yy+433) and (training_chaplain<6) and (global.chapter_name!="Space Wolves") and (global.chapter_name!="Iron Hands"){cooldown=8000;training_chaplain+=1;scr_income();}
-        if (mouse_y>=yy+435) and (mouse_y<yy+452) and (training_psyker<6) and (string_count("Intolerant",obj_ini.strin2)=0){cooldown=8000;training_psyker+=1;scr_income();}
+        if (mouse_y>=yy+355) and (mouse_y<yy+373) and (recruiting<5) and (obj_controller.gene_seed>0) and (obj_ini.doomed==0) and (string_count("|",obj_controller.recruiting_worlds)>0) and (obj_controller.penitent==0){
+            cooldown=8000;
```
Comment (OH296) [inline objects/obj_controller/Mouse_50.gml:327]: trello reminder to change obj_controller.recruiting_worlds to an array?

**Review by OH296 — COMMENTED** [review]
All looks fine to me basically, would say it's likely to cause errors if merged before squads due to changes of how units are handled but not sure on that.

File: objects/obj_controller/Mouse_50.gml [hunk 7]
```diff
@@ -112,96 +139,119 @@ if (menu=12) and (cooldown<=0) and (penitorium>0){
...
-            var i;i=menu_artifact;
+            var i=menu_artifact;
 
-            if (menu_artifact=fest_display) then fest_display=0;
+            if (menu_artifact==fest_display) then fest_display=0;
 
             if (string_count("Daemon",obj_ini.artifact_tags[i])>0){
                 if (obj_ini.artifact_sid[i]>=500){
-                    var oops;oops=floor(random(100))+1;
+                    var demonSummonChance=irandom(100)+1;
```
Comment (GalacticChimp) [inline objects/obj_controller/Mouse_50.gml:222]: finally this makes sense 👍

**Review by GalacticChimp — APPROVED** [review]

**Pull request merged**